src/export_data.py

```diff
@@ -0,0 +1,145 @@
+    neighborhood_kernel = ee.Kernel.square(kernel_radius)
+    # vertical and horizontal kernels used in calculating valid pixels
+    vertical_kernel = ee.Kernel.rectangle(xRadius=1,
+                                          yRadius=kernel_radius,
+                                          units='pixels')
+    horizontal_kernel = ee.Kernel.rectangle(xRadius=kernel_radius,
+                                            yRadius=1,
+                                            units='pixels')
+    # Initializing SampleExporter and TaskManager
+    task_manager = TaskManager(verbose=True)
+    sampler = SampleExporter(task_manager, num_samples, num_shards,
+                             neighborhood_kernel, scale, bucket_name,
+                             export_folder)
+    # Initializing imagery classes
+    multispectral = MultiSpectralImagery(params['collections']['multispectral'],
+                                         start_date, end_date,
+                                         bands=params['bands']['multispectral'],
+                                         scale=scale)
+    tropomi = TropomiImagery(params['collections']['tropomi'],
+                             bands=params['bands']['tropomi'])
+    # Wind imagery adds wind bands from the previous 12 hours
+    wind = WindImagery(params['collections']['wind'],
+                       bands=params['bands']['wind'])
+    # DSM Imagery returns the most recent data prior to the tropomi date
+    dsm = DSMImagery(params['collections']['dsm'], scale,
+                     params['bands']['dsm'])
+    # road is an image, returns clipped image cropped to multispectral region
+    road = RoadImagery(params['collections']['road'],
+                       bands=params['bands']['road'])
+
+    print("Stacking bands for %i multispectral images" % len(multispectral))
+    # We loop through Multispectral Images
+    # First 2 images already exported
+    for j in range(2, len(multispectral)):
+
+        multispectral_image = multispectral[j]
+        image_info = multispectral_image.getInfo()
+        image_id = image_info['properties']['productionID']
+        print("Exporting for Multispectral N:%i, %s" % (j, image_id))
+        # ImageCollection of images that contain all_bands
+        # ImageCollection is pre-filtered so that alll images have at least
+        # num_samples valid pixels
+        stacked_bands_images = stack_bands_from_imagery(multispectral_image,
+                                                        wind, dsm, road,
+                                                        tropomi,
+                                                        vertical_kernel,
+                                                        horizontal_kernel,
+                                                        num_samples,
+                                                        all_bands, scale)
+        size = stacked_bands_images.size().getInfo()
+        listed = stacked_bands_images.toList(size)
```

Comment:
This shouldn't be done in the client since your adding a barrier to the computation (i.e. list out the number of images), is there any reason you **aren't** forming a feature collection from the join between images and samples and sharding that? That sounds a lot simpler and should execute faster.

Reply:
I tried it and I had computation limit errors. But I'm not sure if it was due to flattening the collection of collections